libraries/SDFS/src/SDFSFormatter.h

```diff
@@ -0,0 +1,403 @@
+/*
+ SDFSFormatter.cpp - Formatter for SdFat SD cards
+ Copyright (c) 2019 Earle F. Philhower, III. All rights reserved.
+
+ A C++ implementation of the SdFat/examples/SdFormatter sketch:
+ | Copyright (c) 2011-2018 Bill Greiman
+
+ This library is free software; you can redistribute it and/or
+ modify it under the terms of the GNU Lesser General Public
+ License as published by the Free Software Foundation; either
+ version 2.1 of the License, or (at your option) any later version.
+
+ This library is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ Lesser General Public License for more details.
+
+ You should have received a copy of the GNU Lesser General Public
+ License along with this library; if not, write to the Free Software
+ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#ifndef _SDFSFORMATTER_H
+#define _SDFSFORMATTER_H
+
+#include "SDFS.h"
+#include <FS.h>
+
+namespace sdfs {
+
+class SDFSFormatter {
+private:
+    // Taken from main FS object
+    sdfat::Sd2Card *card;
+    sdfat::cache_t *cache;
+
+    uint32_t cardSizeBlocks;
+    uint32_t cardCapacityMB;
+
+
+    // MBR information
+    uint8_t partType;
+    uint32_t relSector;
+    uint32_t partSize;
+
+    // Fake disk geometry
+    uint8_t numberOfHeads;
+    uint8_t sectorsPerTrack;
+
+    // FAT parameters
+    uint16_t reservedSectors;
+    uint8_t sectorsPerCluster;
+    uint32_t fatStart;
+    uint32_t fatSize;
+    uint32_t dataStart;
+
+    uint8_t writeCache(uint32_t lbn) {
+        return card->writeBlock(lbn, cache->data);
+    }
+
+    void clearCache(uint8_t addSig) {
+        memset(cache, 0, sizeof(*cache));
+        if (addSig) {
+            cache->mbr.mbrSig0 = sdfat::BOOTSIG0;
+            cache->mbr.mbrSig1 = sdfat::BOOTSIG1;
+        }
+    }
+
+    bool clearFatDir(uint32_t bgn, uint32_t count) {
+        clearCache(false);
+        if (!card->writeStart(bgn, count)) {
+            DEBUGV("SDFS: Clear FAT/DIR writeStart failed");
+            return false;
+        }
+        for (uint32_t i = 0; i < count; i++) {
+            if ((i & 0XFF) == 0) {
```

Comment:
We really need a better mechanism than this. I still think optimistic_yield should get an overhaul to yield every x ms.

Reply:
Need a new API defined or you'll break existing code.  Could you think of a way of using your templated-timeout for something here?  I think it would fit, but I can't quite get it...

Reply:
Instantiate a periodic and call yield() on timeout. It's kind of the inverse of the canonical usage.

Reply:
I've put in what I think will give a 5ms delay between yields.  Can you check the new code in this area?  Seems too simple to work, but I didn't have any problem with running it.

Reply:
Looks correct :smiley: